### PR TITLE
Describe Architecture of the server

### PR DESCRIPTION
**Pull request by DeltaEvo**

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -13,3 +13,133 @@ Branch are usualy merged with a fast-forward merge (sometimes with a rebase befo
...
+Each actions can be transformed using functions called `transform` functions. A transform function is a function that take
+an action and return a new action
+
+Transform functions are called from the `transform_action` function inside `src/index.js`
+
+### View
+
+The view is implmemented using functions called observers.
+
+An observer is a function that take a `context` object as parameter, this object contains a `client` ([Client](https://github.com/PrismarineJS/node-minecraft-protocol/blob/master/docs/API.md#mcclientisserverversioncustompackets)), an `events` ([EventEmitter](https://nodejs.org/api/events.html#events_class_eventemitter)) a `dispatch` function (`dispatch(type, payload)`) and a `world` object
```
Comment (Mathis-Paul-62) [inline CONTRIBUTING.md:83]: An observer is a function that takes

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -13,3 +13,133 @@ Branch are usualy merged with a fast-forward merge (sometimes with a rebase befo
...
+```
+
+Each actions can be transformed using functions called `transform` functions. A transform function is a function that take
+an action and return a new action
+
+Transform functions are called from the `transform_action` function inside `src/index.js`
+
+### View
+
+The view is implmemented using functions called observers.
```
Comment (Mathis-Paul-62) [inline CONTRIBUTING.md:81]: implemented

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -13,3 +13,133 @@ Branch are usualy merged with a fast-forward merge (sometimes with a rebase befo
...
+
+An observer is a function that take a `context` object as parameter, this object contains a `client` ([Client](https://github.com/PrismarineJS/node-minecraft-protocol/blob/master/docs/API.md#mcclientisserverversioncustompackets)), an `events` ([EventEmitter](https://nodejs.org/api/events.html#events_class_eventemitter)) a `dispatch` function (`dispatch(type, payload)`) and a `world` object
+
+The `client` is the connection with the minecraft client, you can send and receive packets with it
+
+The `events` is an event channel, a `state` event is emitted when the [State](#State) change, you can use it for custom events
+
+The `dispatch` function is used to create [Actions](#Actions)
+
+The `world` is an imutable object see [World](#World)
```
Comment (Mathis-Paul-62) [inline CONTRIBUTING.md:91]: immutable

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -13,3 +13,133 @@ Branch are usualy merged with a fast-forward merge (sometimes with a rebase befo
...
+		...
+	})
+}
+```
+
+Observers are called from the `observe_client` function inside `src/index.js`
+
+## World
+
+The World is an immutable object that hold the ressources of a world, Merchants, Entity IDs, ...
```
Comment (Mathis-Paul-62) [inline CONTRIBUTING.md:105]: that holds

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -13,3 +13,133 @@ Branch are usualy merged with a fast-forward merge (sometimes with a rebase befo
...
+- Seamless World Game Server: Each server handle a part of the map
+- Responsibility-Oriented Game Server: The server is split in different servers that handle different operations, like Microservices
+
+### Distributed Network Connections
+
+We are already handling all clients individualy, so we only need to synchronise between servers the `events` in the (World)[#World] and Chunks events
+
+### Seamless World Game Server
+
+We need to be able to send a player between server, we only have one [State](#State) so we only need to send it to the new server
```
Comment (Mathis-Paul-62) [inline CONTRIBUTING.md:137]: between servers

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -13,3 +13,133 @@ Branch are usualy merged with a fast-forward merge (sometimes with a rebase befo
...
+We also need to synchronise `events` in the (World)[#World] and Chunks events for the chunk boundaries
+
+### Responsibility-Oriented Game Server
+
+We need to synchronise (Actions)[#Actions] between servers and run only a subset of each observers + reducers on each servers
```
Comment (Mathis-Paul-62) [inline CONTRIBUTING.md:142]: each server

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -13,3 +13,133 @@ Branch are usualy merged with a fast-forward merge (sometimes with a rebase befo
 This codebase try to follow the [Functionnal Programming](https://en.wikipedia.org/wiki/Functional_programming) Paradigm,
 if you don't know this paradigm you can still contribute. Try to learn it while developping your contribution it's not required
 to be perfect. We will help you during the review to make it perfect :)
+
+# Architecture
+
+## State Management Pattern
+
+The Code Base follow a State Management Pattern (like [Redux](https://redux.js.org/introduction/core-concepts) or [Vuex](https://vuex.vuejs.org/#what-is-a-state-management-pattern))
```
Comment (Mathis-Paul-62) [inline CONTRIBUTING.md:21]: follows

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -13,3 +13,133 @@ Branch are usualy merged with a fast-forward merge (sometimes with a rebase befo
...
+
+The code is breaked into 3 parts:
```
Comment (Mathis-Paul-62) [inline CONTRIBUTING.md:23]: BROKEN

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -13,3 +13,133 @@ Branch are usualy merged with a fast-forward merge (sometimes with a rebase befo
...
+
+Theses part communicates with a "one-way data flow"
+
+![Flow](media/flow.svg)
+
+### State
+
+Each minecraft client have it's own state, this state is updated using a function called a reducer.
+
+A reducer is just a function that take the state and an action as argument and return the next state of the client
```
Comment (Mathis-Paul-62) [inline CONTRIBUTING.md:37]: takes

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -13,3 +13,133 @@ Branch are usualy merged with a fast-forward merge (sometimes with a rebase befo
...
+  }
+  return state
+}
+```
+
+Theses reducers are called from the `reduce_state` function inside `src/index.js`
+
+### Actions
+
+Actions are javascript objects in the form `{ type: String, payload: any }` that describe what happened
```
Comment (Mathis-Paul-62) [inline CONTRIBUTING.md:55]: that describes

**Review by Sceat — APPROVED** [review]

**Pull request merged**